combats uploader race condition

craigsteffen opened the pull request:
Putting in a CONDITIONAL wait to combat a race condition in the distributed file system.  It first checks to see if the file exists. If it doesn't, it waits .1 seconds, if it still doesn't, it waits 1.0 seconds.  Then it raises high holy hell about the file not existing.  

So if it does indeed exist, we've only incurred the cost of running a single existence check and an if, so it will only do the wait if it already HAS hit the race condition.

craigsteffen committed:
if-->wait race condition

Putting in a CONDITIONAL wait to combat a race condition in the distributed file system.  It first checks to see if the file exists. If it doesn't, it waits .1 seconds, if it still doesn't, it waits 1.0 seconds.  Then it raises high holy hell about the file not existing.  

So if it does indeed exist, we've only incurred the cost of running a single existence check and an if, so it will only do the wait if it already HAS hit the race condition.

craigsteffen committed:
print -> logging.exception

doing error reporting correctly via logging.exception (instead of print() )

craigsteffen committed:
update for 0.7.3

Proper log message for 0.7.3

robkooper commented on uploader/uploader.py:
```diff
@@ -30,6 +31,13 @@ def run(self):
+                if not os.path.exists(file):
+                    time.sleep(1)
+                    if not os.path.exists(file):
+                        logging.exception(f"ERROR!  File {file} does not exist for uploader even after a wait!")
```
Use logging.exception when you want to print the stack trace for the exception. Since there is no exception at this point, no need to do this. Additionally we will print the raised exception on line 44 so no need to print it twice, just the raise is good enough.

craigsteffen replied:
@robkooper 

I'm trying to parse this out.  

I *think* you're saying that when the checks for the file are exhausted, it's not appropriate to call "logging.exception".  But are you saying to leave the "raise ValueError"?  

I just want to make sure that if the uploader fails because the file it's been told to upload doesn't exist, then what file it's looking for is explicitly stated and that "does not exist", so that the user knows it's not a network problem or a queue problem or something.

asaxton replied:
Maybe a point of clarity: The `logging.exception()` is being called in the `try:` block that started on line 30. At line 44, no exception as been raised that `logging.exception()` can use in its massaging.

As a loose rule, `logging.exception()` should only be called in `except:` blocks.

The `except:` block is where the program skips to when an exception is raised and thus there is a traceback availbile that `logging.exception()` can "read" and populate the logging message with. Notice this is what is happening between line 51 and 56

craigsteffen replied:
@robkooper I used the github interface to apply the patch.  

Once I got the file back, I don't think it was right.  So I spliced the two trees together the way I think you were thinking, which I agree with; it's much more compact.  And I removed the logging.exception but left the raise ValueError.  

Let me know if that matches what you were thinking.  If you think this looks Ok, we should probably exercise the uploader both normally, without the race condition, and then artificially trip it so that it definitely fires, so to make sure it flows correctly.

craigsteffen replied:
@asaxton 

Ah, thanks, that helps a lot.  

Unfortunately I uploaded a new commit just as you posted that comment.  Could you glance at the new commit?  The patch tool in github kind of mangled the test loops and the new commit cleans them up.  

Also I _think_ I've fixed my original misplaced "except" statement.

asaxton replied:
Everything looks correct to me.

robkooper commented on uploader/uploader.py:
```diff
@@ -30,6 +31,13 @@ def run(self):
             data = json.loads(self.body)
             file = os.path.join("/output", data['cdr_output'])
             logging.debug(f"Uploading data for {data['cog_id']} from {file}")
+            if not os.path.exists(file):
```
To prevent all the if if if statements, additionally this makes it easy to try once twice, n times.
```suggestion
            counter = 0
            while not os.path.exists(file):
                counter = counter + 1
                if counter > 2: # maybe make a variable above
                    raise ValueError(f"File {file} does not exist for uploader!!!")  
                time.sleep(1)
```

craigsteffen replied:
Ok, yeah, this makes sense.  I'll commit this, and then make the parameters (wait time and wait count) variables that are set in the function so they're easy to adjust.

craigsteffen committed:
if if -> while not w/ counter

Replacing tree if ifs (for os.path.exists for the target) with a while(not exists) loop with a counter.  Makes it more flexible and more compact.

Co-authored-by: Rob Kooper <kooper@illinois.edu>

craigsteffen committed:
clean up file check if tree

I don't think the github patch applier got the final version of the
code right.  I think this is what Rob intended with his revised while
loop replacing the couple of ifs I had initially written.
